### PR TITLE
Crosschain resolver

### PR DESCRIPTION
**Pull request by makoto**
L1 Owned Resolver that can fetch verifiable data against L2 chains.

(Builds are currently failing until it merges https://github.com/ensdomains/evmgateway/pull/6)

### CODE REVIEW

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+pragma solidity ^0.8.17;
+
+import {EVMFetcher} from '@ensdomains/evm-verifier/contracts/EVMFetcher.sol';
+import {EVMFetchTarget} from '@ensdomains/evm-verifier/contracts/EVMFetchTarget.sol';
+import {IEVMVerifier} from '@ensdomains/evm-verifier/contracts/IEVMVerifier.sol';
+
+contract L1Resolver is EVMFetchTarget {
+    using EVMFetcher for EVMFetcher.EVMFetchRequest;
+    uint256 private constant COIN_TYPE_ETH = 60;
+    IEVMVerifier verifier;
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:11]: These can be immutable.

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    function addr(bytes32 node) public view returns (address) {
+        EVMFetcher
+            .newFetchRequest(verifier, target)
+            .getStatic(1)
+            .element(node)
+            .getDynamic(3)
+            .ref(0)
+            .element(node)
+            .element(COIN_TYPE_ETH)
+            .fetch(this.addrCallback.selector, ''); // recordVersions
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:33]: Indentation can make this easier to read:
```suggestion
        EVMFetcher.newFetchRequest(verifier, target)
            .getStatic(1)
              .element(node)
            .getDynamic(3)
              .ref(0)
              .element(node)
              .element(COIN_TYPE_ETH)
            .fetch(this.addrCallback.selector, ''); // recordVersions
```

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    ) public view returns (bytes memory) {
+        EVMFetcher
+            .newFetchRequest(verifier, target)
+            .getStatic(1)
+            .element(node)
+            .getDynamic(3)
+            .ref(0)
+            .element(node)
+            .element(coinType)
+            .fetch(this.addrCoinTypeCallback.selector, ''); // recordVersions
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:61]:
```suggestion
        EVMFetcher.newFetchRequest(verifier, target)
            .getStatic(1)
              .element(node)
            .getDynamic(3)
              .ref(0)
              .element(node)
              .element(coinType)
            .fetch(this.addrCoinTypeCallback.selector, ''); // recordVersions
```

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    ) public view returns (string memory) {
+        EVMFetcher
+            .newFetchRequest(verifier, target)
+            .getStatic(1)
+            .element(node)
+            .getDynamic(11)
+            .ref(0)
+            .element(node)
+            .element(key)
+            .fetch(this.textCallback.selector, '');
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:89]:
```suggestion
        EVMFetcher.newFetchRequest(verifier, target)
            .getStatic(1)
              .element(node)
            .getDynamic(11)
              .ref(0)
              .element(node)
              .element(key)
            .fetch(this.textCallback.selector, '');
```

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+     */
+    function contenthash(bytes32 node) public view returns (bytes memory) {
+        EVMFetcher
+            .newFetchRequest(verifier, target)
+            .getStatic(1)
+            .element(node)
+            .getDynamic(4)
+            .ref(0)
+            .element(node)
+            .fetch(this.contenthashCallback.selector, '');
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:112]:
```suggestion
        EVMFetcher.newFetchRequest(verifier, target)
            .getStatic(1)
              .element(node)
            .getDynamic(4)
              .ref(0)
              .element(node)
            .fetch(this.contenthashCallback.selector, '');
```

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+     * @param coinType The cointype to query
+     * @return The associated address.
+     */
+    function addr(
+        bytes32 node,
+        uint256 coinType
+    ) public view returns (bytes memory) {
+        EVMFetcher
+            .newFetchRequest(verifier, target)
+            .getStatic(1)
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:55]: The various magic numbers here should be constants.

File: crosschain-resolver/scripts/remote.ts
Comment (Arachnid) [inline crosschain-resolver/scripts/remote.ts:1]: Is this needed?

File: l1-verifier/scripts/test.js [hunk 7]
```diff
@@ -10,36 +10,43 @@ async function main() {
   const server = ganache.server(options);
   console.log('Starting server');
   const port = await new Promise((resolve, reject) => {
-    server.listen(0, async (err) => {
```
Comment (Arachnid) [inline l1-verifier/scripts/test.js:13]: Can you revert these changes?

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: l1-verifier/scripts/test.js (hunk 7 above)
Comment (makoto) [reply l1-verifier/scripts/test.js:13]: When I switched from `fork` to `spawn` the option to pass RPC_PORT dynamically wasn't working so I decided to hardcode the port. Would hardcoding port number cause any issues?

Comment (Arachnid) [reply l1-verifier/scripts/test.js]: Yes; there's no guarantee a fixed port number is available. Why did you need to switch from `fork` to `spawn`?

Comment (makoto) [reply l1-verifier/scripts/test.js]: `fork` takes the path to run the command and you were specifying the relative path where `spawn` takes the actual command that runs in the context of the environment

https://nodejs.org/docs/latest-v18.x/api/child_process.html#child_processforkmodulepath-args-options

Comment (Arachnid) [reply l1-verifier/scripts/test.js]: Okay; was it not working? It worked for me and in the build.

Comment (makoto) [reply l1-verifier/scripts/test.js]: Which build? This change is already in the `main` branch as part of this PR https://github.com/ensdomains/evmgateway/pull/6

Comment (Arachnid) [reply l1-verifier/scripts/test.js]: I missed that it was in the other pr. Can you please revert it? I still don't understand why you needed to make this change in the first place.

File: crosschain-resolver/README.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    ) public view returns (bytes memory) {
+        EVMFetcher
+            .newFetchRequest(verifier, target)
+            .getStatic(1)  // storage_slot of recordVersions
+            .element(node)
+            .getDynamic(3) // storage_slot of versionable_addresses
+            .ref(0)        // Referencing the result of `.getStatic(0)`
+            .element(node)
+            .element(coinType)
+            .fetch(this.addrCoinTypeCallback.selector, ''); // recordVersions
```
Comment (Arachnid) [inline crosschain-resolver/README.md:48]:
```suggestion
        EVMFetcher.newFetchRequest(verifier, target)
            .getStatic(1)  // storage_slot of recordVersions
              .element(node)
            .getDynamic(3) // storage_slot of versionable_addresses
              .ref(0)        // Referencing the result of `.getStatic(0)`
              .element(node)
              .element(coinType)
            .fetch(this.addrCoinTypeCallback.selector, ''); // recordVersions
```

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+contract L1Resolver is EVMFetchTarget {
+    using EVMFetcher for EVMFetcher.EVMFetchRequest;
+    uint256 private constant COIN_TYPE_ETH = 60;
+    IEVMVerifier immutable verifier;
+    address target;
+    uint256 constant RECORD_VERSIONS_SLOT = 1;
+    uint256 constant VERSINABLE_ADDRESSES_SLOT = 3;
+    uint256 constant VERSINABLE_HASHES_SLOT = 4;
+    uint256 constant VERSINABLE_TEXTS_SLOT = 11;
+    uint8 constant RECORD_VERSIONS_REF = 0;
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:17]: Can you reorder these to put constants, immutables, etc, all together?

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+import {EVMFetcher} from '@ensdomains/evm-verifier/contracts/EVMFetcher.sol';
+import {EVMFetchTarget} from '@ensdomains/evm-verifier/contracts/EVMFetchTarget.sol';
+import {IEVMVerifier} from '@ensdomains/evm-verifier/contracts/IEVMVerifier.sol';
+
+contract L1Resolver is EVMFetchTarget {
+    using EVMFetcher for EVMFetcher.EVMFetchRequest;
+    uint256 private constant COIN_TYPE_ETH = 60;
+    IEVMVerifier immutable verifier;
+    address target;
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:12]: I believe this can also be immutable.

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    function contenthashCallback(
+        bytes[] memory values,
+        bytes memory
+    ) public pure returns (bytes memory) {
+        return values[1];
+    }
+
+    function bytesToAddress(
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:123]: I believe you can just cast the `bytes` to a `bytes20` rather than using assembly.

File: crosschain-resolver/deploy_l2/01_l2resolver .ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+  const {deploy} = deployments;
+
+  const {deployer} = await getNamedAccounts();
+  console.log({deployer})
+  await deploy('OwnedResolver', {
+    from: deployer,
+    args: [],
+    log: true,
+    gasLimit: 5000000,
+    gas: 1000000000000
```
Comment (Arachnid) [inline crosschain-resolver/deploy_l2/01_l2resolver .ts:15]: Is it necessary to supply these?

File: crosschain-resolver/scripts/remote.ts
Comment (Arachnid) [inline crosschain-resolver/scripts/remote.ts:1]: Is this file needed for anything?

File: crosschain-resolver/test/testResolver.ts [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    const contenthash = '0xe3010170122029f2d17be6139079dc48696d1f582a8530eb9805b561eda517e22a892c7e3f1f'
+    await l2contract.clearRecords(node)
+    await l2contract.setContenthash(node, contenthash)
+    await provider.send("evm_mine", []);
+    const result = await l2contract.contenthash(node)
+    expect(result).to.equal(contenthash);
+    const result2 = await target.contenthash(node, { enableCcipRead: true })
+    expect(result2).to.equal(contenthash);
+  })
+});
```
Comment (Arachnid) [inline crosschain-resolver/test/testResolver.ts:135]: Some tests that demonstrate the behaviour when records don't exist would be valuable.

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+contract L1Resolver is EVMFetchTarget {
+    using EVMFetcher for EVMFetcher.EVMFetchRequest;
+    IEVMVerifier immutable verifier;
+    address immutable target;
+    uint256 constant COIN_TYPE_ETH = 60;
+    uint256 constant RECORD_VERSIONS_SLOT = 1;
+    uint256 constant VERSINABLE_ADDRESSES_SLOT = 3;
+    uint256 constant VERSINABLE_HASHES_SLOT = 4;
+    uint256 constant VERSINABLE_TEXTS_SLOT = 11;
+    uint8   constant RECORD_VERSIONS_REF = 0;
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:17]: It doesn't really make sense for this to be a constant.

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: crosschain-resolver/contracts/L1Resolver.sol (hunk 14 above)
Comment (makoto) [reply crosschain-resolver/contracts/L1Resolver.sol:17]: You mean just RECORD_VERSIONS_REF or all other constants?

Comment (Arachnid) [reply crosschain-resolver/contracts/L1Resolver.sol:17]: Just that one.

**Comment by makoto** [comment]
@Arachnid PTAL

**Review by Arachnid — APPROVED** [review]
